File: code/algo/getoffer/go/2/bst_right_view.go
package main

func rightSideView(root *TreeNode) []int {
	ans := []int{}
	if root == nil {
		return ans
	}
	var dfs func(node *TreeNode, depth int)
	dfs = func(node *TreeNode, depth int) {
		if node == nil {
			return
		}

		if len(ans) == depth {
			ans = append(ans, node.Val)
		}
		dfs(node.Right, depth+1)
		dfs(node.Left, depth+1)
	}
	dfs(root, 0)
	return ans
}

File: code/algo/getoffer/c++/2/bst_right_view.cpp
#include <vector>

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
  std::vector<int> rightSideView(TreeNode *root) {
    std::vector<int> ans;
    if (!root) {
      return ans;
    }
    dfs(root, 0, ans);
    return ans;
  }

private:
  void dfs(TreeNode *node, int depth, std::vector<int> &ans) {
    if (!node) {
      return;
    }
    if (depth == ans.size()) {
      ans.push_back(node->val);
    }
    dfs(node->right, depth + 1, ans);
    dfs(node->left, depth + 1, ans);
  }
};
